### ISSUE TITLE
Draggable and resizable image layers

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#299]
## 🚀 Feature
It would be nice to be able to drag and resize image layers using their `translate`, `scale`, and potentially `rotate` options. This would allow people doing image registration or template matching to attempt their best guess manually. Image border functionality #298 and the interactive box like functionality from the shapes layer could be reused to make this possible

**Comment by sofroniewn** [comment]
One further thought - should this functionality actually be on the base layer and so accessible to all layers?

**Comment by jni** [comment]
Yes, I think it makes little sense to make it part of the image layer only. This way you could select e.g. an image and its markers layer together and transform them together.

**Comment by sofroniewn** [comment]
Ok making it a generic property of the base layer sounds good, as it makes sense for all our different layer types. Note that it is different then selecting all the markers and moving them - as that  would affect their data coordinates, whereas this would only be affecting the layer translate and scale properties (i.e. the scene graph stuff). It's possible to apply these transformation stuff to our `svg` generation too, so that it remains consistent.

I'm not sure though about allowing multiple layers to be transformed at the same time. Right now if you have multiple layers selected the controls panel on the left goes away and I imagine this being an option in that panel. @jni do you feel strongly that simultaneously manipulating multiple layers needs to be supported?

**Comment by jni** [comment]
> Note that it is different then selecting all the markers and moving them - as that would affect their data coordinates, whereas this would only be affecting the layer translate and scale properties (i.e. the scene graph stuff).

Yes, I understand this. Incidentally it would be nice to be able to get at this data programmatically, ie, load two images, move one or the other, then grab the transform data. Ditto for *setting* the transform — instant visualization of registrations!

> do you feel strongly that simultaneously manipulating multiple layers needs to be supported?

Well, if it is a property of the base-layer, then maybe this interaction can live on the layers/layerslist directly? I am only going by intuition, but I think it might be very useful to select multiple layers and drag them together. One might annotate an image, then add a new image and want to translate the image and annotations for consistency. If you have to translate them separately, you lose the original relationship.

An intermediate option would be the ability to apply the transformation of one layer to another.

**Comment by sofroniewn** [comment]
The programatic API for translation and scaling is already there, just set the `layer.translate` and `layer.scale` properties to a 2-tuple and a float respectively. I don't think it would be too much work to expose the `rotate` property from vispy, but I need to look into that. It should be pretty easy to add such an example to our current examples. We could also expose a convenience method that took an affine transform specified how `skimage` specifies them and applies it to a `Layer`.

My concern with adjusting multiple layers simultaneously was more of a UI consistency one (all of our operations only apply to one layer at a time right now) rather than an implementation one, but I agree this would be a useful feature. We can get the implementation working for a single layer first and then discuss further

**Comment by jni** [comment]
> I don't think it would be too much work to expose the rotate property from vispy, but I need to look into that.

It looks like we would have to switch from `STTransform` to `MatrixTransform`, which I think is the right move anyway.

http://vispy.org/visuals.html#module-vispy.visuals.transforms

> We can get the implementation working for a single layer first and then discuss further

:+1:

**Comment by sofroniewn** [comment]
sounds good - bumping this to our `0.2` milestone as per discussion in #298

**Comment by sofroniewn** [comment]
@AhmetCanSolak just gave this feature request a +1 I think, correct?

**Comment by AhmetCanSolak** [comment]
Absolutely, i think this is valuable for overall layers experience.

**Comment by GenevieveBuckley** [comment]
This would be really relevant in our lab, we need to do plenty of image registration between different imaging modalities. Currently we're using control point matching, which works ok. 

We'd be happy to test things, if that helps.

**Comment by GenevieveBuckley** [comment]
This feature request was discussed in our last lab meeting. Some very useful points were made, which I'll summarize here:

1. **When rotating images, it's best if there's some way to allow the user to change the center of rotation as needed within each image.** It's easy to see why - any time the rotation center is far away from the image regions you are trying to match up, things tend to swing around wildly (and it's harder still if there is a translation involved at the same time).

2. **Having the ability to add constraints to the transformation was also seen as something we needed.** A common expected use case is:
- Click and drag images manually to roughly align them
- Identify some spot that is the same in both images, click that to add it as a constraint to the transformation.
- Click and drag the images some more (now subject to some constraints on the way they can move relative to one another)
- Identify some other spot that is the same in both images, and add that as a second constraint.
- ... either the user reaches a point where they are happy with the (semi-)manual alignment, or they add enough constraint points that the transformation can be calculated exactly.

**Comment by sofroniewn** [comment]
@GenevieveBuckley thanks for those added details. You might also be interest in #763 where we are discussing adding a physical coordinate system. After we've done that it might be a good time to think about this feature again and as @jni works on #763 it will probably be good to keep use-cases like this in mind

**Comment by psobolewskiPhD** [comment]
I think with https://github.com/napari/napari/pull/3611 and https://github.com/napari/napari/pull/4894 this issue is a good ways towards being addressed.
Remaining:
- GUI: The new mode is available via keyboard shortcut (default `2`) but not otherwise exposed in the GUI.
- the new mode doesn't work in 3D, just freezes manipulation of the rendered image.
anything else?

**Comment by psobolewskiPhD** [comment]
Transform mode is now part of the GUI as of napari 0.5.0
https://github.com/napari/napari/pull/6794

It does work for 3D layers but can only be activated in 2D mode -- you can use orthogonal views tough, but can get:
`UserWarning: Non-orthogonal slicing is being requested, but is not fully supported. Data is displayed without applying an out-of-slice rotation or shear component.!`

I'm going to close this, as non-orthogonal slicing is a different can of works.
